Faster bn254 syscalls

cryptopapi997 opened the pull request:
Inspired by https://github.com/anza-xyz/solana-sdk/pull/408 , we can ~steal~ adapt these ideas to optimize the bn254 syscalls as well, since those all overwrite the full memory buffer as well. For the compression syscalls that return a fixed size array, we can utilize the same `MaybeUninit` trick. For the addition, multiplication, and pairing syscalls we can alloc a vec once at the beginning instead of allocing an array and then converting that to a vec. This results in the following CU diffs:

| operation      | Δ CU |
| ----------- | ----------- |
| add | -27 |
| mul | -27 |
| pairing | -13 |
| g1_compress | -4 |
| g2_compress | -8 |
| g1_decompress | -8 |
| g2_decompress | -22 |

Note: I didn't apply this to the deprecated bn254 syscalls, but can do as well.
Note 2: Technically these optimizations could be applied inside the actual syscall implementation too instead of the wrapper functions interfacing with them for a couple extra CUs. However, this seemed more low risk, and idk if a SIMD would apply in that case so I went with this for now.

joncinque reviewed (commented):
Thanks for your contribution, looks great to me! Just the one comment, @samkim-crypto can you also take a look?

joncinque commented:
Noting for anyone watching, that the `Vec::with_capacity` -> `as_mut_ptr()` -> `set_len()` flow is very well demonstrated here: https://doc.rust-lang.org/std/vec/struct.Vec.html#method.as_mut_ptr

samkim-crypto reviewed (commented):
Just one small clarification below, but everything looks fine to me!

cryptopapi997 commented on bn254/src/compression.rs:
```diff
@@ -345,152 +345,156 @@ mod target_arch {
-        match result {
-            0 => Ok(result_buffer),
-            _ => Err(AltBn128CompressionError::UnexpectedError),
+                result_buffer.as_mut_ptr() as *mut u8,
```
No, it returns `*mut [u8; OUTPUT_LENGTH]` bc `as_mut_ptr` just returns `*mut T` and result_buffer is a fixed size byte array. If we remove the cast, rust complains since it's a different type than the expected `*mut u8` (even though we know these are the same here)

cryptopapi997 commented:
I changed `&mut result_buffer as *mut _ as *mut u8` to `as_mut_ptr` bc we introduced dealing with the raw bytes of `vec`s in some of the code. For these, those two operations are very different, the former will give a pointer to the beginning of the struct (i.e. the length field) whereas the latter will give a pointer to the first element (what we want). For the raw `MaybeUninit` arrays you're right this switch is technically not needed, but imo lowers the possibility of someone seeing the pointer cast being done differently, trying to unify them, and messing up the vec stuff & just felt a bit cleaner overall.

> Also, you didn't make this change in alt_bn128_g1_compress_be, alt_bn128_g1_decompress_be, alt_bn128_g2_compress_be, alt_bn128_g2_compress_le, alt_bn128_pairing

I mentioned this in my PR description, `alt_bn128_pairing` is deprecated, so I left it as is. For the other compression syscalls I did [make the change](https://github.com/anza-xyz/solana-sdk/blob/master/bn254/src/compression.rs#L423-L424) using `core::mem::MaybeUninit`, unless you're thinking of a different optimization I missed.
